### DISCUSSION TITLE
Zustand state is cleared during Next.js deployment

### DISCUSSION THREAD

**Discussion by uki1014** [vercel/next.js#55941] [Show and tell] [answered]
I have encountered an event in a production Next.js application, where the Zustand state was cleared during deployment.
At that time, the request log showed that the state managed by useState was not cleared.

Next.js seems to use the [Automatic Updates](https://nextjs.org/docs/app/building-your-application/deploying#automatic-updates) mechanism to automatically modernize client-side JavaScript without reloading if the latest version exists.
(I don't understand much about the details regarding Auto updates in Next.js...)

Does this destroy objects referenced by libraries like Zustand that are running outside of React?

- version
  - Next.js: 12.3.1
  - Zustand: 4.3.9

**Comment by icyJoseph** [comment]
This sounds weird. 

I guess your zustand store is setup in _app? And it gets props calculated either by gIP, or GSP, or GSSP? Then maybe when the JSON for a page is pre-fetched, and then you navigate into the page, the props that drive the zustand store get reset?

**Reply by uki1014** [reply]
@icyJoseph
Thank you for your comment.

Zustand store is initialized within a React component in one of the pages, not in _app. The usage is the same as for CSR, and the screen does not use either gIP, GSP, or GSSP.

Next.js is supposed to automatically update when it finds the latest deployments, but does it have a mechanism to keep the state?

For example, with Zustand, it seems that internally the state is held as just a JavaScript object, so is it possible that the state in memory is destroyed when the JavaScript being loaded is replaced by a new one by deployment(Automatic Updates)?

**Reply by icyJoseph** [reply]
Right, but is it doing the automatic update, while you are on that page already?

Or is it more like, you navigate to another page and come back, and now your stuff is gone?

**Reply by uki1014** [reply]
Since this event occurred on the end-user's client, we are judging from the contents of its log. (The Zustand state, which should be there, was empty, and the useState state was included in the request without any problem.)
And the Next.js web server was deploying at the time this request occurred.

Judging from the log, there is only one page, and it seems to occur when switching tabs within that page.
The end-user opened tabA and change Zustand's state, then opened tabB and chagne useState's state.
Then a request is made to the API server combining Zustand's state and useState's state.

**Reply by icyJoseph** [reply]
mmm this is getting even more strange - so the React state is persisted, but not the Zustand store?

The tabs inside this page do not trigger URL changes right?

**Reply by uki1014** [reply]
> so the React state is persisted, but not the Zustand store?

I assume from the following documentation that when fetching new JavaScript files after deployment, Next.js can use the official React API (e.g useState) to maintain local state, but the state of external libraries such as Zustand is cleared I assumed that the state of external libraries such as Zustand would be cleared.

[https://nextjs.org/docs/architecture/fast-refresh#limitations](https://nextjs.org/docs/architecture/fast-refresh#limitations:~:text=Fast%20Refresh%20tries%20to%20preserve%20local%20React%20state%20in%20the%20component%20you%27re%20editing%2C%20but%20only%20if%20it%27s%20safe%20to%20do%20so.%20Here%27s%20a%20few%20reasons%20why%20you%20might%20see%20local%20state%20being%20reset%20on%20every%20edit%20to%20a%20file%3A)

> The tabs inside this page do not trigger URL changes right?

The URLs are the same, but queryparams are changed to express the state of switching tabs.

**Reply by icyJoseph** [reply]
Mmm but Fast Refresh is a dev mode thing, isn't it? it should impact production builds.

And how is the URL navigation done then? Via next/router or next/link in some fashion? It does say in Automatic Updates, that a next/link behaves as an `a` tag during Automatic Updates, but why would the React state persist 🤔 unless local storage or some other API is at play here? Puzzling

**Reply by uki1014** [reply]
@icyJoseph 
We investigated and found that it was an application-side defect that occurs only under complex conditions, so this discussion is closed.
The Next.js deployment mechanism was never affected.

Thank you for your reply. :pray:

**Reply by icyJoseph** [reply, marked as answer]
@uki1014 ah, of course. That's what I was after initially, but at _app level. 

Good that you got it!
